src/serverlist.cpp

```diff
@@ -631,6 +635,10 @@ void CServerListManager::Remove ( const CHostAddress& InetAddr )
  - SERVER external to DIRECTORY (list entry external IP same LAN)
    - CLIENT internal to SERVER (CLM     same IP as list entry external IP): use "internal" fields (local LAN address)
    - CLIENT external to SERVER (CLM not same IP as list entry external IP): use "external" fields (public IP address from protocol)
+
+ Finally, when retrieving the entry for the directory itself life is more complicated still.  It's easiest just to return "0"
+ and allow the client connect dialogue instead to use the IP and Port from which the list was received.
```

Comment:
I haven't tried it, but I think the connect dialog should work independent of this PR as it works based on index 0 without looking at the address at all.

It's other protocol consumers (such as Jamulus Explorer) which break without this change.

Reply:
That means it's open to question where the fix should be made.

If Jamulus Explorer gets the address/port right with this fix, I'd rather leave the change in place and remove the work around in the client connect dialogue.  The above is meant as a reminder that there's code there needing investigation...  Maybe it says the wrong thing, though.

Reply:
I think this PR is correct and should be merged because 3.9.0 unintentionally changed protocol details. We don't have a formal specification of the protocol, so the Jamulus implementation effectively dictates the details.

Removing the logic in the client is not a trivial thing, IMO. If we remove it, *all* servers would have to generate a real, non-empty index 0, which is close to impossible... :)
Moving the logic from the GUI to the client makes sense, of course, and should be transparent to do.

Reply:
> Removing the logic in the client is not a trivial thing, IMO.

I didn't say the _client_ shouldn't do it.  Actually, I think it should be in `src/protocol.cpp` when receiving the list.  That way any use would get the right value.  However, it _definitely_ shouldn't be in the UI.  That layer should _NOT_ be required for correct behaviour.  The RPC client, for example, _should_ be able to retrieve the server list and connect to any entry.